ikura/shim/src/dock/rollkit.rs

```diff
@@ -1,122 +1,363 @@
+                        hex::encode(&block_hash),
+                        err,
+                    )));
+                }
+            };
+            let blob_id = BlobId {
+                block_number,
+                extrinsic_index,
+                data_hash,
+            };
+            info!("blob landed: {blob_id}");
+            response.ids.push(blob_id.into());
         }
-        // TODO:
-        Ok(0)
+        Ok(Response::new(response))
+    }
+
+    async fn get_proofs(
+        &self,
+        request: Request<GetProofsRequest>,
+    ) -> Result<Response<GetProofsResponse>, Status> {
+        // TODO: implement
+        // https://github.com/thrumdev/blobs/issues/257
+        let GetProofsRequest { ids, .. } = request.into_inner();
+        let response = GetProofsResponse {
+            proofs: ids
+                .into_iter()
+                .map(|_| pbda::Proof { value: vec![] })
+                .collect(),
+        };
+        Ok(Response::new(response))
+    }
+
+    async fn validate(
+        &self,
+        request: Request<ValidateRequest>,
+    ) -> Result<Response<ValidateResponse>, Status> {
+        // TODO: implement
+        // https://github.com/thrumdev/blobs/issues/257
+        let ValidateRequest { ids, .. } = request.into_inner();
+        let response = ValidateResponse {
+            results: ids.into_iter().map(|_| true).collect(),
+        };
+        Ok(Response::new(response))
+    }
+
+    async fn commit(
+        &self,
+        request: Request<CommitRequest>,
+    ) -> Result<Response<CommitResponse>, Status> {
+        // TODO: implement
+        // https://github.com/thrumdev/blobs/issues/257
+        let CommitRequest { blobs, .. } = request.into_inner();
+        let response = CommitResponse {
+            commitments: blobs
+                .into_iter()
+                .map(|_| pbda::Commitment { value: vec![] })
+                .collect(),
+        };
+        Ok(Response::new(response))
     }
 }
 
-/// Parses the namespace from a given string encoded as hex.
-///
-/// Note that rollkit uses arbitrary length namespaces, but ikura uses 16-byte namespaces. For
-/// now, we would silently truncate or pad the namespace to 16 bytes.
-fn parse_namespace(namespace: &str) -> anyhow::Result<ikura_nmt::Namespace> {
-    use ikura_nmt::NS_ID_SIZE;
-    let namespace_bytes = hex::decode(namespace)?;
-    if namespace_bytes.len() != NS_ID_SIZE {
-        debug!(
-            "The namespace '{}' is not {} bytes long. Resizing...",
-            namespace, NS_ID_SIZE
-        );
+impl RollkitDock {
+    /// Returns the namespace to be used, either from the request or from the configuration.
+    ///
+    /// If the namespace is not provided in the request, it will use the namespace from the
+    /// configuration.
+    fn obtain_namespace(
+        &self,
+        supplied_ns: Option<pbda::Namespace>,
+    ) -> Result<ikura_nmt::Namespace, Status> {
+        Ok(match supplied_ns {
```

Comment:
To avoid wrapping everything into an `Ok(...)` another approach could probably be:

```rust
        match supplied_ns {
            Some(pbda::Namespace {
                value: raw_namespace_bytes,
            }) => {
                let raw_namespace_bytes = raw_namespace_bytes.as_slice();
                if raw_namespace_bytes.len() != 16 {
                    return Err(Status::invalid_argument("namespace must be 16 bytes long"));
                }
                let mut namespace = [0u8; 16];
                namespace.copy_from_slice(raw_namespace_bytes);
                Ok(ikura_nmt::Namespace::from_raw_bytes(namespace))
            }
            None => self.namespace
                .and_then(|namespace| Ok(namespace.clone()))
                .ok_or(Status::invalid_argument("namespace must be provided"))
        }
```

Reply:
Looks terser and like a good suggestion - @gabriele-0201 this is a good opportunity to practice on Graphite. Instead of pausing this PR and updating it, create a new PR based on this stack which applies that change as a refactor.

Reply:
oh and yeah `namespace.clone()` is not necessary.